algorithmic_efficiency/checkpoint_utils.py

```diff
@@ -0,0 +1,162 @@
+
+USE_PYTORCH_DDP, RANK, DEVICE, N_GPUS = pytorch_setup()
+CheckpointReturn = Tuple[spec.OptimizerState,
+                         spec.ParameterContainer,
+                         spec.ModelAuxiliaryState,
+                         dict,
+                         list,
+                         int,
+                         int]
+
+
+def maybe_restore_checkpoint(framework: str,
+                             optimizer_state: spec.OptimizerState,
+                             model_params: spec.ParameterContainer,
+                             model_state: spec.ModelAuxiliaryState,
+                             train_state: dict,
+                             eval_results: list,
+                             global_step: int,
+                             preemption_count: int,
+                             checkpoint_dir: str) -> CheckpointReturn:
+  if framework == 'jax':
+    opt_state, opt_update_fn = optimizer_state
+  else:
+    opt_state, opt_update_fn = optimizer_state, None
+
+  uninitialized_global_step = -1
+  uninitialized_preemption_count = -1
+  checkpoint_state = dict(
+      model_params=model_params,
+      optimizer_state=opt_state,
+      model_state=model_state,
+      train_state=train_state,
+      eval_results=None,
+      global_step=uninitialized_global_step,
+      preemption_count=uninitialized_preemption_count)
+
+  if framework == 'jax':
+    latest_ckpt = flax_checkpoints.restore_checkpoint(
+        checkpoint_dir, target=checkpoint_state)
+    save_path = os.path.join(checkpoint_dir,
+                             'checkpoint_' + str(latest_ckpt['global_step']))
+  else:
+    latest_ckpt = checkpoint_state
+    save_path = os.path.join(checkpoint_dir, 'checkpoint')
+    if os.path.exists(save_path):
+      latest_ckpt = torch.load(save_path, map_location=DEVICE)
+
+  # Load_latest_checkpoint() will return checkpoint_state if
+  # checkpoint_dir does not exist or if it exists and contains no checkpoints.
+  found_checkpoint = (latest_ckpt['global_step'] != uninitialized_global_step)
+
+  if not found_checkpoint:
+    return (optimizer_state,
+            model_params,
+            model_state,
+            train_state,
+            eval_results,
+            global_step,
+            preemption_count)
+
+  # If there's the latest checkpoint in the checkpoint_dir, restore from that.
+  if framework == 'jax':
+    checkpoint_state = replicate_checkpoint(
+        latest_ckpt,
+        pytree_keys=[
+            'optimizer_state',
+            'model_params',
+            'model_state',
+        ])
+    checkpoint_state['optimizer_state'] = (checkpoint_state['optimizer_state'],
+                                           opt_update_fn)
+    checkpoint_state['eval_results'] = [
+        (value, key) for key, value in latest_ckpt['eval_results'].items()
+    ]
+
+  else:
+    checkpoint_state = latest_ckpt
+    model_params.load_state_dict(checkpoint_state['model_params'])
```

Comment:
Do we need a `map_location` argument here (even when the state dict was saved from GPU and ist loaded to GPU, it might be necessary or at least safer to specify the correct device when using DDP)?

Reply:
Please see my below comment, regarding the device map location.